components/brave_wallet/browser/simulation_response_parser.cc

```diff
@@ -0,0 +1,492 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/strings/string_number_conversions.h"
+#include "brave/components/brave_wallet/browser/simulation_response_parser.h"
+#include "brave/components/brave_wallet/browser/simulation_responses.h"
+
+namespace brave_wallet {
+std::vector<mojom::BlowfishWarningPtr> ParseWarnings(
+    const std::vector<simulation_responses::Warning>& values) {
+  std::vector<mojom::BlowfishWarningPtr> warnings;
+  for (const auto& warning : values) {
+    warnings.push_back(mojom::BlowfishWarning::New(
+        warning.severity, warning.kind, warning.message));
+  }
+  return warnings;
+}
+
+absl::optional<std::string> ParseNullableString(const base::Value& value) {
+  if (!value.is_none()) {
+    return value.GetString();
+  }
+
+  return absl::nullopt;
+}
+
+namespace evm {
+mojom::BlowfishPricePtr ParsePrice(const base::Value& value) {
+  if (!value.is_none()) {
+    const auto& price_value =
+        simulation_responses::Price::FromValue(value.GetDict());
+    return mojom::BlowfishPrice::New(price_value->source,
+                                     price_value->updated_at,
+                                     price_value->dollar_value_per_token);
+  }
+
+  return nullptr;
+}
+
+mojom::BlowfishEVMContractPtr ParseContract(
+    const simulation_responses::EVMContract& value) {
+  return mojom::BlowfishEVMContract::New(value.address, value.kind);
+}
+
+mojom::BlowfishEVMAmountPtr ParseAmount(
+    const simulation_responses::EVMAmount& value) {
+  return mojom::BlowfishEVMAmount::New(value.before, value.after);
+}
+
+mojom::BlowfishEVMAssetPtr ParseAsset(
+    const simulation_responses::EVMAsset& value) {
+  auto asset = mojom::BlowfishEVMAsset::New();
+  asset->address = value.address;
+  asset->symbol = value.symbol;
+  asset->name = value.name;
+
+  if (!base::StringToInt(value.decimals, &asset->decimals)) {
+    return nullptr;
+  }
+
+  asset->verified = value.verified;
+
+  if (value.lists) {
+    asset->lists = *value.lists;
+  } else {
+    asset->lists = {};
+  }
+
+  asset->image_url = ParseNullableString(value.image_url);
+  asset->price = ParsePrice(value.price);
+  return asset;
+}
+
+mojom::BlowfishEVMStateChangeRawInfoPtr ParseStateChangeRawInfo(
+    const simulation_responses::EVMStateChangeRawInfo& value) {
+  auto raw_info = mojom::BlowfishEVMStateChangeRawInfo::New();
+  raw_info->kind = value.kind;
+
+  if (value.kind == "ERC20_TRANSFER") {
+    auto data_value = simulation_responses::ERC20TransferData::FromValue(
+        value.data.GetDict());
```

Comment:
is it guranteed that value.data is a dict?

Reply:
We're now doing an early-exit if `value.data` is not a dict. Also added tests for the same.

c37ccc62a37e718ade35982b6c5b8a90e3b2c8f9